lib/use_case/lodge_assessment.rb

```diff
@@ -145,27 +142,42 @@ def find_related_rrn(wrapper_hash)
     end
 
     def save_new_address_id(assessment)
-      new_address_id = assessment.address_id
-
-      if new_address_id.nil?
-        new_address_id = default_address_id(assessment)
-      elsif new_address_id.start_with?("UPRN-")
-        # TODO: Maybe in the future, prevent assessors from lodging non existing UPRNs
-        uprn = new_address_id[5..-1].to_i.to_s
-        uprn_search_result = @address_base_search_gateway.search_by_uprn(uprn)
-        new_address_id = default_address_id(assessment) if uprn_search_result
-          .empty?
-      end
+      new_address_id = get_new_address_id(assessment)
+      source =
+        get_assessments_address_id_source(
+          current_address_id: assessment.address_id,
+          new_address_id: new_address_id,
+        )
 
       @assessments_address_id_gateway.send_to_db(
         {
           assessment_id: assessment.assessment_id,
           address_id: new_address_id,
-          source: "lodgement",
+          source: source,
         },
       )
     end
 
+    def get_new_address_id(assessment)
+      if assessment.address_id&.start_with?("UPRN-")
+        # TODO: Maybe in the future, prevent assessors from lodging non existing UPRNs
+        uprn = assessment.address_id[5..-1]
+        return assessment.address_id if address_base_has_uprn?(uprn)
+      elsif assessment.address_id&.start_with?("RRN-")
+        related_assessment_id = assessment.address_id[4..-1]
+        begin
+          related_assessment =
+            @assessments_address_id_gateway.fetch(related_assessment_id)
+        rescue StandardError => e
+          related_assessment = nil
+        end
+
+        return related_assessment[:address_id] if related_assessment
+      end
+
+      default_address_id(assessment)
```

Comment:
I found the execution flow to be more difficult to understand. Before it was clear the default address was invoked only when assessment.address_id was nil. Because we do not check nil explicitly, we have to use conditional method calls and a conditional return statement.

What about checking for nil, then UPRN, RRN or else? (which would explicitly catch everything, LPRN can still be recorded via migration).